### ISSUE TITLE
Extending NetworkService API to allow for Chaining

### ISSUE THREAD

**Issue by edwarnicke** [ligato/networkservicemesh#531]
This takes inspiration for Istio VirtualService.

Currently, NetworkService CRs look roughly like:

```yaml
kind: NetworkService
metadata:
  name: secure-intranet-connectivity
spec:
  payload: IP
```

Basically, each NetworkService has a name and a payload type.

If more than one NetworkServiceEndpoint provides a NetworkService, in response to a request for a Connection to a Network Service from a NetworkServiceClient, the nsmd round robins among them.

There are lots of cases however where it is helpful to have more control, to be able to Chain together NetworkServiceEndpoints (NSEs) to provide a single Network Service:

![network service extensions for chaining](https://user-images.githubusercontent.com/1455208/48921483-b7597180-ee65-11e8-978c-f54bc01dac69.png)

This issue is about putting together a basic extension to NetworkService to allow us to build such chains.

Each Network Service Endpoint when it [advertise itself](https://github.com/ligato/networkservicemesh/blob/master/controlplane/pkg/apis/registry/registry.proto#L50) includes a [set of labels](https://github.com/ligato/networkservicemesh/blob/master/controlplane/pkg/apis/registry/registry.proto#L13).  For the purpose of this discussion, lets call those ```destinationLabels```.

Each Network Service Client with it [requests a Connection to a NetworkService](https://github.com/ligato/networkservicemesh/blob/master/controlplane/pkg/apis/local/networkservice/networkservice.proto#L34) includes a [set of labels](https://github.com/ligato/networkservicemesh/blob/master/controlplane/pkg/apis/local/connection/connection.proto#L44).  For the purposes of this discussion, lets call those ```sourceLabels```.

We can extend the Network Service to be more like:

```yaml
kind: NetworkService
metadata:
  name: secure-intranet-connectivity
spec:
  payload: IP
  matches:
  - match:
    sourceSelector:
        app: firewall
    route:
      - destination:
          destinationSelector:
            app: vpn-gateway
  - match:
    route:
      - destination:
          destinationSelector:
            app:firewall
```

Now when selecting the Network Service Endpoint to connect an NSC to, the nsmd goes through the 'matches' until one matches.

So when a Client Pod, which has no sourceLabel "app=firewall" requests a connection, it will fail the first match.  The second match, which has no sourceSelector will match *any* connection, so it matches, and the Client Pod is connected to a Network Service Endpoint that has a destinationLabel "app=firewall".

The Firewall Pod itself has to request "secure-intranet-connectivity".  It makes such a request, but it has in its sourceLabels "app=firewall".  When the nsmd seeks to select a NetworkServiceEndpoint for the Firewall Pods request, it will match the first match, and select a Network Service Endpoint with "app=vpn-gateway".

This allows a simple flexible way to make selection decisions for L2/L3 Connections.

Let's extend this with one further by allowing multiple 'destinations' in a route with weights:

```yaml
kind: NetworkService
metadata:
  name: secure-intranet-connectivity
spec:
  payload: IP
  matches:
  - match:
    sourceSelector:
        app: firewall
    route:
      - destination:
          destinationSelector:
            weight: 90
            app: vpn-gateway
            version: v1.0
      - destination:
          destiinationSelector:
            weight: 10
            app: vpn-gateway
            version: v2.0
  - match:
    route:
      - destination:
          destinationSelector:
            app:firewall
```

In the above example, 10% of the traffic that matches sourceSelector "app=firewall" is sent to a vpn-gateway with a label indicationg its version is v2.0, the other 90% is sent to a vpn-gateway with a label indicating its version is v1.0.

**Comment by edwarnicke** [comment]
## Pointers for adding Extending the Network Service API to support Chaining

Please note, this Issue is for extending the API, another will be written for providing pointers on *implementing* selecting in the nsmd.

You will need to extend the [NetworkServiceSpec](https://github.com/ligato/networkservicemesh/blob/master/k8s/pkg/apis/networkservice/v1/types.go#L27).

You will need to extend the [NetworkService message](https://github.com/ligato/networkservicemesh/blob/master/controlplane/pkg/apis/registry/registry.proto#L17)

You will need to enhance the [translation of the NetworkServiceSpec to the NetworkServiceMessage](https://github.com/ligato/networkservicemesh/blob/master/k8s/pkg/registryserver/nse.go#L154) in the RegistryServer.

**Comment by matrohon** [comment]
I can't wait to see this in action :1st_place_medal:

**Comment by nickolaev** [comment]
So when we do the 90%/10% split we do than on connection selection level. I think it will be more correct to state that "10% of the newly established connections will go to vpn-gateway v2.0". I mean it is not traffic load balancing/splitting, it is just new connection selection that is affected.

**Comment by edwarnicke** [comment]
@nickolaev You are correct.  90%/10% split is on connections, not traffic.

**Comment by nickolaev** [comment]
I'd suggest that we don't include `weight` in the `destinationSelector`, i.e. get something like this:

```yaml
kind: NetworkService
metadata:
  name: secure-intranet-connectivity
spec:
  payload: IP
  matches:
  - match:
    sourceSelector:
        app: firewall
    route:
      - destination:
          weight: 90                      # weight is `destination` property now
          destinationSelector:
            app: vpn-gateway
            version: v1.0
      - destination:
          weight: 10
          destiinationSelector:
            app: vpn-gateway
            version: v2.0
  - match:
    route:
      - destination:
          destinationSelector:
            app:firewall
```

This make `destinationSelector` a `labelSelector` type, and leaves `weight` as a generic `destination` property.

**Comment by edwarnicke** [comment]
Closed in #537
